Provide UI for entering admin password when an admin password is set

grzesiek2010 opened the pull request:
Closes #4683 
Closes #4695  
Closes #4699

#### What has been done to verify that this works as intended?
I tested the fix manually and added/improved tests.

#### Why is this the best possible solution? Were any other approaches considered?
The biggest change I implemented and based my changes on are states.

Before if admin password was set we were asked to enter it before opening settings so there were only two states:
- settings in normal mode
- settings in admin mode

Now we are able to open settings without entering password (we ill be asked for it if we try to access protected settings).
This required implementing new states:
- LOCKED, // Admin password is set but not entered
- UNLOCKED, // Admin password is set and entered
- NOT_PROTECTED // Admin password is not set

#### How does this change affect users? Describe intentional changes to behavior and behavior that could have accidentally been affected by code changes. In other words, what are the regression risks?
All settings should be tested. I would be rather focused on visibility/accessibility because the functionality itself shouldn't be changed.

#### Do we need any specific form for testing your changes? If so, please attach one.
No.

#### Does this change require updates to documentation? If so, please file an issue [here]( https://github.com/getodk/docs/issues/new) and include the link below.
No.

#### Before submitting this PR, please make sure you have:
- [x] run `./gradlew checkAll` and confirmed all checks still pass OR confirm CircleCI build passes and run `./gradlew connectedDebugAndroidTest` locally.
- [x] verified that any code or assets from external sources are properly credited in comments and/or in the [about file](https://github.com/getodk/collect/blob/master/collect_app/src/main/assets/open_source_licenses.html).
- [x] verified that any new UI elements use theme colors. [UI Components Style guidelines](https://github.com/getodk/collect/blob/master/CONTRIBUTING.md#ui-components-style-guidelines)

grzesiek2010 committed:
Renamed icons

grzesiek2010 committed:
Added new icons

grzesiek2010 committed:
Removed redundant code prom ProjectSettingsDialogTest

grzesiek2010 committed:
Converted AdminPasswordDialogFragment to kotlin and imporved it

grzesiek2010 committed:
Converted ChangeAdminPasswordDialog to kotlin and improved it

grzesiek2010 committed:
Disable protected settings if admin password is set

grzesiek2010 committed:
Improved ProjectPreferencesFragment adding consts for preference keys

grzesiek2010 committed:
Display 'locked' icon if admin pasword is set

grzesiek2010 committed:
Ask for admin password after clicking on protected prefs

grzesiek2010 committed:
Replaced interface in AdminPasswordDialogFragment with viewmodel

grzesiek2010 committed:
Refresh toolbar icons when password is entered

grzesiek2010 committed:
Do not allow setting admin password if it's set but not entered

grzesiek2010 committed:
React to setting admin password

grzesiek2010 committed:
Fixed showing/removing prefs

grzesiek2010 committed:
Removed redundant tests

grzesiek2010 committed:
Manage state of settings using viewmodel

grzesiek2010 committed:
Added titles for menu items

grzesiek2010 committed:
Code and naming improvements

grzesiek2010 committed:
Apply access control if admin password is not set

grzesiek2010 committed:
Improved MainMenuAccessPreferencesTest and converted to kotlin

grzesiek2010 committed:
Converted MainMenuAccessPreferencesFragment to kotlin

grzesiek2010 committed:
Implemented SingleEventLiveData to avoid receiving dupicated events after entering or setting password

grzesiek2010 committed:
Improved ProjectPreferencesViewModel using factory

grzesiek2010 committed:
Converted IdentityPreferencesFragmentFragmentTest to kotlin and improved it

grzesiek2010 committed:
Convereted IdentityPreferencesFragment to kotlin

grzesiek2010 committed:
Converted FormManagementPreferencesFragmentTest to kotlin and fixed it

grzesiek2010 committed:
Converted UserInterfacePreferencesFragmentTest to kotlin and fixed it

grzesiek2010 committed:
Fixed and improved ProjectPreferencesFragmentTest

grzesiek2010 committed:
Fixed removing disabled prefs from the main settings screen

lognaturel commented:
Looking really good!

On Android 7.1, I'm not seeing any visual indication that the items in the protected section are disabled when an admin password is set and hasn't been entered. Maybe it's an issue specific to this Android version?

Now that I see it in action, I wonder whether not applying access controls when the admin password has been entered is too confusing. Doesn't have to be resolved in this PR but something to think about and consider.

I also think that at some point (doesn't even have to be this release) we should give some behavior to tapping on the open lock. It feels unexpected that it doesn't do anything. Maybe it could show a confirmation dialog to lock settings and apply access controls or something like that.

When there's no admin password set, the access controls are applied which is great. It looks like when access control is changed to hide something, it refreshes immediately. But when access control is changed to show something, it does not.
- Access control > User settings > Uncheck server
- Go back to Project settings, Server isn't there 👍
- Access control > User settings > Check server
- Go back to Project settings, Server still isn't there
- Rotate the screen, Server appears

grzesiek2010 committed:
Fixed ConfigureWithQRCodeTest

grzesiek2010 committed:
Improved refreshing prefs

grzesiek2010 commented:
> On Android 7.1, I'm not seeing any visual indication that the items in the protected section are disabled when an admin password is set and hasn't been entered. Maybe it's an issue specific to this Android version?

No it's not not implemented. I can't make those prefs disabled because then they are not clickable so I would just need to set the colors somehow programmatically but does that make sense. Isn't the padlock icon enough? 

> When there's no admin password set, the access controls are applied which is great. It looks like when access control is changed to hide something, it refreshes immediately. But when access control is changed to show something, it does not.

Now it should be fixed.

I still want to add more automated tests and I will do that tomorrow, but if you think the implementation is good enough we can ask Kasia to test it.

lognaturel commented:
> I can't make those prefs disabled because then they are not clickable so I would just need to set the colors somehow programmatically but does that make sense. Isn't the padlock icon enough?

Ah, I see. It feels a lot weirder to me than I expected. That is, if I have a password set, when I tap on "Set Admin Password", I don't expect to be prompted for the existing password because there's no indication that it will do anything other than let me set the password. It's also very disorienting that the screen refreshes and scrolls to the top but doesn't seem to have done the action I expected.

Maybe a simple answer is to hide the whole protected section if there's an admin password and it hasn't been entered? I'm not sure why I didn't just do that in the first place... 😭 

> ask Kasia to test it.

Yes, that would be good. I still do want to do another pass looking at a couple quick implementation things but I'll mark as `needs testing`. We can decide whether the changes described above should go in this PR or a follow-on.

lognaturel commented:
> I'm not sure why I didn't just do that in the first place... 😭

Going back to the Figma file reminded me that I was worried that the lock wouldn't be intuitive to folks who already have an admin password set and are upgrading. They might be looking for admin settings and feel really lost so showing some familiar settings but disabled seemed like a way to guide upgrading users towards what they're looking for. 

I'm not sure, though, what do you think? Maybe the lock is clear enough? Would it be straightforward to add a tip like [this](https://www.figma.com/file/qOwdGX7QgNOAzQ10TVxH6h/ODK-Collect-settings?node-id=29%3A611)?

grzesiek2010 committed:
Improved managing keys in AccessControlPreferencesFragment

grzesiek2010 committed:
Improved managing keys in ProjectPreferencesFragment

grzesiek2010 committed:
Removed redundant val

grzesiek2010 committed:
Converted BasePreferencesFragment to kotlin

grzesiek2010 commented:
> Maybe the lock is clear enough?

I guess it should be enough so i would rather just hide those protected prefs.

The second scenario would require creating our custom preference to display a message like that, it's doable but I'm not sure if it makes sense.

grzesiek2010 committed:
Removed duplicated code responsible for displaying new preferences (navigation)

grzesiek2010 committed:
Moved initializing the first preferences scren from activity to xml

grzesiek2010 committed:
Removed redundant setTitle()

grzesiek2010 committed:
Add protected setting to hide Project display category

grzesiek2010 committed:
Improved ProjectPreferencesFragmentTest

seadowg commented on collect_app/src/main/java/org/odk/collect/android/preferences/screens/ProjectPreferencesFragment.kt:
```diff
@@ -39,90 +45,167 @@ class ProjectPreferencesFragment :
+            this,
+            { isPasswordSet: Boolean ->
+                if (isPasswordSet) {
+                    projectPreferencesViewModel.setStateUnlocked()
```
I think we can potentially simplify this. Right now, we have two ViewModels interacting via observations from Fragments. 

Instead, could we share the `ProjectPreferencesViewModel` between the `ProjectPreferencesFragment` and the `AdminPasswordDialogFragment`. I'm imagining the ViewModel could have an `attemptPassword` method that checks the password and then sets the locked/unlocked state based on that. The locked state could then also be a `LiveData` exposed by the `ProjectPreferencesViewModel` and the fragment could decide how it renders based on an observation of that. 

For showing Toasts etc I think exposing a separate `LiveData` for "errors" or "failed attempts" could work. This could use a `LiveData<Consumable>` instead of a `SingleEventLiveData`. I prefer this as it allows observers to make their own decision about what to do with the "event" rather than changing the semantics of `LiveData`. 

Does that make sense @grzesiek2010 @lognaturel?

grzesiek2010 replied:
Like this https://github.com/getodk/collect/pull/4687/commits/a2ba0a754b27b5bcaf58cf8ed78ab48478612dab?
Yeah it makes the code a bit cleaner. I forgot about `LiveData<Consumable>` I can use it but I'm still not sure if I like it more than that `SingleEventLiveData`...

seadowg replied:
Yup, liking the new structure a lot more! I'm wondering if we could get rid of the `Consumable` entirely, but I think there is some awkwardness with the prefs lifecycle and `LiveData` that we'd need to get right, so probably a tweak for another day!

grzesiek2010 committed:
Cleaned displayPreferences() method

grzesiek2010 committed:
Share ProjectPreferencesViewModel between fragments/dialog fragments and replace SingleEventLiveData with Consumable

kkrawczyk123 commented:
**Tested with success!**
Verified on Androids: 5.1, 8.1, 10 and 11.

Verified cases:
- light mode/dark mode
- screen rotation
- minimizing app
- set admin password
- edit admin password
- access settings when admin password set
- create a project through qr code when admin password set 
- create a project through qr code when settings hidden
- reset settings when admin password selected 
- incorrect admin password toast
- show and hide settings when locked/unlocked
- set admin password on few projects
- hide/unhide project display
- create project without admin password set but with hidden settings
- hide all settings options


@lognaturel I have noticed that now when general settings and admin settings are gathered together I am able to jump to admin settings and edit them in the middle of filling form. I haven't noticed anything especially weird as those changes are not updated immediately and I have to leave form (ex. turn on and off moving backwards) but I want to make sure that you are aware of that.

lognaturel commented:
I like the latest change! Great to see that QA didn't reveal anything serious. I'd like to see this merged before we make any final behavior updates.

> I am able to jump to admin settings and edit them in the middle of filling form

Ah yes, thanks for highlighting this, @kkrawczyk123. It's not really ideal but since you're not finding any problems with it, I think we go with it for now.

grzesiek2010 commented:
@lognaturel so what about those protected prefs? Can I just hide them?

lognaturel commented:
> what about those protected prefs?

Let's leave for now and keep thinking about the best approach. I'll file a separate issue. It will be isolated enough that I think it's fine to add after/during regression testing.

grzesiek2010 committed:
Improved tests

grzesiek2010 committed:
Added ProjectPreferencesViewModelTest

lognaturel commented:
I've filed #4699. Like I've said, fine to do after you've merged. Or if one of the two ideas I am leaning towards sounds good to you and is straightforward to implement, you could do it here. Up to you!

grzesiek2010 committed:
Hide protected settings if admin password is set

grzesiek2010 committed:
Use setValue() instead of postValue() in ProjectPreferencesViewModel

grzesiek2010 committed:
Update preferences visibility instead of adding/removing them for better user experience

grzesiek2010 committed:
Fixed tests

grzesiek2010 committed:
Fixed DisabledPreferencesRemover

grzesiek2010 committed:
Added tests

grzesiek2010 committed:
Removed redundant keys

grzesiek2010 committed:
Fixed updating prefs

grzesiek2010 commented:
I've just added a fix for #4699 to this pr (https://github.com/getodk/collect/pull/4687/commits/747ad31a2409c8b701cb38950735d21bdd2b7f04) and also added a nice improvement I think (showing/hiding prefs instead of removing them and adding again https://github.com/getodk/collect/pull/4687/commits/baf2f05febdb34d798ee96ac197a022a6df7cbc1) what makes the experience much better @kkrawczyk123 also noticed that the refreshing was just ugly.

@kkrawczyk123 please take a look at it when you are free.

grzesiek2010 committed:
Fixed tests

kkrawczyk123 commented:
@grzesiek2010 I have noticed weird behavior when followed scenario:
1. Open form
2. Jump to Settings
3. Change ex. font size
4. Try to unlock protected settings
After input admin password they do not get unlocked, see gif:
![ezgif com-gif-maker (6)](https://user-images.githubusercontent.com/33342488/125964535-47e5f727-00f1-4e43-89a0-c11a20cc1295.gif)

When followed the same scenario but without any edit on the settings side, it works well.

lognaturel commented:
And this does NOT happen when launching settings from the main menu?

I think we should merge, start regression testing (Monday I guess), and explore in parallel.

lognaturel commented:
I checked and it's not related to launching settings from form entry, it happens regardless. So if you ~change non-protected settings~ navigate to any other settings screen and then unlock, the UI doesn't reflect the unlock and there doesn't seem to be any way to make it do so. Maybe a listener is getting lost or something?

That's serious enough that it should be fixed before merge.

grzesiek2010 commented:
I'm going to investigate it

grzesiek2010 committed:
Fixed updating prefs

grzesiek2010 commented:
This small change fixes the issue https://github.com/getodk/collect/pull/4687/commits/3e2fac0dc77d612fe83b26ab048f503bce213c29
Don't ask my why because I'm still wondering what was wrong...
I debugged the project and everything was called properly so @lognaturel 

> Maybe a listener is getting lost or something?

it wasn't the case.

The same method I moved down in `onViewCreated` (`updatePreferencesVisibility()`) is called either way for example in the described case after entering password so why does that change in `onViewCreated` do the job???

lognaturel commented:
Wowww... that is... interesting. Going to look at it a little bit more to see if it might be hiding some other weird behavior. Thanks for getting a fix in so fast!

lognaturel commented:
I can't really explain why order mattered in that case. ~But I think it would be preferable to entirely remove the `onViewCreated` implementation. The effect it has now is that on view creation `updatePreferencesVisibility` is called twice. There will always be a state from the viewmodel that will not have been consumed once the view is created so you can rely on the observation to do the correct refresh.~ No no, that won't work on rotation. It /is/ weird that the refresh happens twice when the view is first created but let's go with it for now.

@kkrawczyk123 please verify that set of steps again on master and file an issue if you still see anything unusual!
